enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,164 @@
+
+3.  Images for each CLI. The CLI in-cluster manager would provide an extract mechanism for each CLI image, similar to how we currently 
+`oc adm release extract --command oc` and `oc image extract`.   A Custom Resource for each CLI would provide the in-cluster CLI manager information about each CLI and its image.
+In disconnected, the images will be available through a mirrored local registry.  The logic for `oc image extract` can be re-used and extended for extracting the CLIs.  CLIs would
+be managed by users through extending the oc commands we currently have to extract `oc` and `openshift-install` binaries from the release payload.  This would provide the function we need, but looking to the future there is an
+upstream effort we can use to provide an in-cluster CLI manager while also providing a mechanism that the Kubernetes community as a whole can utilize.  That brings us to option 4.
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew, artifacts will be packaged within an image.  Each cluster will have an index of CLI images.  Mirrored local registries will provide binaries to disconnected environments.  
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to this: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* CLI manager will not create or update the ClusterCLI Custom Resources, those will be managed by individual CLIs and their operators. 
+
+### Requirements:
```

Comment:
Should probably make it a clear requirement to support multi-arch?

Reply:
Good point, thx Steve!

Reply:
Yes, binaries will be extracted from images, images will be similar to this: https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel  with multi-arch (images will be built using artifacts from https://mirror.openshift.com/pub/openshift-v4/clients/ )

Reply:
I just don't know if these binaries are published in such a way that tools like `krew` to properly list out tools for the appropriate architecture.

Reply:
Here is an example krew-index plugin.yaml: https://github.com/kubernetes-sigs/krew-index/blob/master/plugins/change-ns.yaml  listing different architectures.  We will use a custom krew-index for our tools, and will install to a user's home directory.